File: src/main/java/sample/design_pattern/TemplateMethod.java
package sample.design_pattern;


interface ITextProcess {
    void printHeader();
    void printBody();
    void printFooter();
}


class HTMLDocument implements ITextProcess {
    private String header;
    private String body;
    private String footer;

    public HTMLDocument(String header, String body, String footer) {
        this.header = header;
        this.body = body;
        this.footer = footer;
    }

    public void printHeader() {
        System.out.println("<head>");
        System.out.println("\t" + header);
        System.out.println("</head>");
    }

    public void printBody() {
        System.out.println("<body>");
        System.out.println("\t" + body);
        System.out.println("</body>");
    }

    public void printFooter() {
        System.out.println(footer);
    }
}

class DocumentOperator {
    private ITextProcess content;

    public DocumentOperator(ITextProcess content) {
        this.content = content;
    }

    public void operate() {
        content.printHeader();
        content.printBody();
        content.printFooter();
    }
}


class TemplateMethod {
    public static void main(String[] args) {
        DocumentOperator client = new DocumentOperator(
            new HTMLDocument("head content", "body content", "footer content"));
        client.operate();
    }
}

File: src/main/scala/sample/design_pattern/TemplateMethod.scala
/*
 * Copyright 2013 SatoShun
 *
 * Abstract Factory Pattern
 */

package sample.design_pattern;


trait ITextProcess {
    def printHeader(): Unit;
    def printBody(): Unit;
    def printFooter(): Unit;
}


class HTMLDocument(val header: String, val body: String, val footer: String) extends ITextProcess {
    def printHeader() = {
        println("<head>")
        println("\t" + header)
        println("</head>")
    }

    def printBody = {
        println("<body>")
        println("\t" + body)
        println("</body>")
    }

    def printFooter = {
        println(footer)
    }
}


class DocumentOperator(val content: ITextProcess) {
    def operate() = {
        content.printHeader()
        content.printBody()
        content.printFooter()
    }
}


object TemplateMethod {
    def main(args: Array[String]) = {
        val client = new DocumentOperator(
            new HTMLDocument("head content", "body content", "footer content"));
        client.operate()
    }
}
